File: 2196-create-binary-tree-from-descriptions/solution.py
from typing import List, Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        val_to_node = {}
        children = set()

        for parent, child, is_left in descriptions:
            # Create nodes dynamically if they don't already exist
            if parent not in val_to_node:
                val_to_node[parent] = TreeNode(parent)
            if child not in val_to_node:
                val_to_node[child] = TreeNode(child)

            # Establish relationship
            if is_left == 1:
                val_to_node[parent].left = val_to_node[child]
            else:
                val_to_node[parent].right = val_to_node[child]

            # Mark this node as a child
            children.add(child)

        # The root is the only parent node that is never a child
        for parent in val_to_node:
            if parent not in children:
                return val_to_node[parent]

        return None

File: 2196-create-binary-tree-from-descriptions/solution.cpp
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        unordered_map<int, TreeNode*> valToNode;
        unordered_set<int> children;

        for (const auto& desc : descriptions) {
            int parentVal = desc[0];
            int childVal = desc[1];
            bool isLeft = desc[2];

            // If parent node doesn't exist yet, create it
            if (valToNode.find(parentVal) == valToNode.end()) {
                valToNode[parentVal] = new TreeNode(parentVal);
            }
            // If child node doesn't exist yet, create it
            if (valToNode.find(childVal) == valToNode.end()) {
                valToNode[childVal] = new TreeNode(childVal);
            }

            // Establish the parent-child connection
            if (isLeft) {
                valToNode[parentVal]->left = valToNode[childVal];
            } else {
                valToNode[parentVal]->right = valToNode[childVal];
            }

            // Record child to find the root later
            children.insert(childVal);
        }

        // The root node has no parent, so it won't be in the children set
        for (const auto& pair : valToNode) {
            int val = pair.first;
            if (children.find(val) == children.end()) {
                return pair.second;
            }
        }

        return nullptr;
    }
};